django/contrib/admin/views/main.py

```diff
@@ -384,3 +389,97 @@ def url_for_result(self, result):
                        args=(quote(pk),),
                        current_app=self.model_admin.admin_site.name)
+
+
+class ForeignKeyJsonView(BaseListView):
+    """
+    View that handles requests from `.AdminForeignKeyWidget`.
+
+    The view only supports HTTP's GET method.
+    """
+    paginate_by = 20
+    admin_site = None
+
+    def get(self, request, *args, **kwargs):
+        """
+        Return a :class:`.django.http.JsonResponse`.
+
+        Example::
+
+            {
+                'results': [
+                    {
+                        'text': "foo",
+                        'id': 123
+                    }
+                ],
+                'more': true
+            }
+
+        """
+        field_id = kwargs.get('field_identifier', request.GET.get('field_identifier', None))
+        if not field_id:
+            raise PermissionDenied('No "field_identifier" provided.')
+        try:
+            self.app_label, self.model_name, self.field_name = field_id.split('.')
+        except ValueError:
+            raise PermissionDenied('Invalid "field_identifier".')
+        self.term = kwargs.get('field_identifier', request.GET.get('term', ''))
+        self.object_list = self.get_queryset()
+        context = self.get_context_data()
+        return JsonResponse({
+            'results': [
+                {
+                    'text': force_text(obj),
+                    'id': obj.pk,
+                }
+                for obj in context['object_list']
+            ],
+            'more': context['page_obj'].has_next()
+        })
+
+    def get_model(self):
+        return self.field.remote_field.model
+
+    model = property(get_model)
+
+    def get_field(self):
+        from django.apps import apps
+        try:
+            original_model = apps.get_model(self.app_label, self.model_name)
+            return original_model._meta.get_field(self.field_name)
+        except (LookupError, FieldDoesNotExist):
+            raise PermissionDenied('Invalid "field_identifier".')
+
+    field = property(get_field)
+
+    def get_queryset(self):
+        """Filter queryset based on the models' ``limit_choices_to`` and search term."""
+        queryset = super(ForeignKeyJsonView, self).get_queryset()
+        queryset = queryset.using(router.db_for_read(self.model))
+        model_field = self.get_field()
+        if hasattr(model_field, 'get_limit_choices_to'):
+            limit_choices_to = model_field.get_limit_choices_to()
+            if limit_choices_to is not None:
+                queryset = queryset.complex_filter(limit_choices_to)
+        search_fields = self.get_search_fields()
+        select = reduce(
+            lambda x, y: Q(**{"%s__icontains" % x: self.term}) | Q(**{"%s__icontains" % y: self.term}),
+            search_fields,
+            Q(**{"%s__icontains" % search_fields.pop(): self.term})
+        )
+        return queryset.filter(select).order_by()
```

Comment:
django admin's [`get_search_results`](https://github.com/django/django/blob/8b1110ddff8ca7f300ccef2a5a91c843cd140cf0/django/contrib/admin/options.py#L877-L908) adds a `distinct` on the queryset in some cases (search-fields over m2m rels). 

But btw, since we get the search fields from the existing model-admin options: 
can't we also use `.get_search_results` directly on the admin-instance of the foreign-key?

And just extend the QuerySet with the result of `limit_choices_to`?